lib/cartopy/crs.py

```diff
@@ -338,43 +339,98 @@ def boundary_distance(xy):
         # Record the positions of all the segment ends
         for i, line_string in enumerate(line_strings):
             first_dist = boundary_distance(line_string.coords[0])
-            thing = _Thing(first_dist, False,
-                           (i, 'first', line_string.coords[0]))
+            thing = _BoundaryPoint(first_dist, False,
+                                   (i, 'first', line_string.coords[0]))
             edge_things.append(thing)
             last_dist = boundary_distance(line_string.coords[-1])
-            thing = _Thing(last_dist, False,
-                           (i, 'last', line_string.coords[-1]))
+            if last_dist == first_dist:
+                # We have a closed line-string. Just offset the last_dist by a
+                # tiny fraction.
+                last_dist += 1e-8  # TODO: Use epsilon.
+                thing = _BoundaryPoint(last_dist, False,
+                                       (i, 'self-closing',
```

Comment:
It seems this concept isn't needed now I've dealt with co-incident _BoundaryPoint sections.